streaming/base/format/base/writer.py

```diff
@@ -298,21 +293,30 @@ def _write_index(self) -> None:
 
     def finish(self) -> None:
         """Finish writing samples."""
-        if self.new_samples:
-            self.flush_shard()
-            self._reset_cache()
-        self._write_index()
-        logger.debug(f'Waiting for all shard files to get uploaded to {self.remote}')
-        self.executor.shutdown(wait=True)
+        if self.event.is_set():
+            # If an error occurred, cancel any outstanding uploads
+            self.cancel_future_jobs()
+        else:
+            # Otherwise finish any remaining uploads and write index, and wait for completion
+            if self.new_samples:
+                self.flush_shard()
+                self._reset_cache()
+            self._write_index()
+            logger.debug(f'Waiting for all shard files to get uploaded to {self.remote}')
+            self.executor.shutdown(wait=True)
+
         if self.remote and not self.keep_local:
             shutil.rmtree(self.local, ignore_errors=True)
 
+        # Final check, in case error occurred in an upload during shutdown
+        if self.event.is_set():
+            raise Exception('One of the threads failed. Check other traceback for more ' +
+                            'details.')
+
     def cancel_future_jobs(self) -> None:
         """Shutting down the executor and cancel all the pending jobs."""
         # Beginning python v3.9, ThreadPoolExecutor.shutdown() has a new parameter `cancel_futures`
         self.executor.shutdown(wait=False, cancel_futures=True)
-        if self.remote and not self.keep_local:
```

Comment:
This is already handled in finish(), which seemed like the right one place to do this cleanup. I think the critical thing for write() to do is cancel future work and error out.

However there's not really a problem with also having this logic here IMHO. It can be left if there's any doubt.

Reply:
I'd prefer to keep this since `cancel_future_jobs` is called in multiple other places -- and respects the `self.keep_local` flag

Reply:
has been resolved with conversation below.